[JENKINS-49406] Evergreen snapshotting data safety system

batmat opened the pull request:
# * LOCKED => refer now to https://github.com/jenkinsci/jep/pull/67 for the official submission**


[JENKINS-49406](https://issues.jenkins-ci.org/browse/JENKINS-49406)

PR to allow a first pass of reviews as I still consider this content an alpha version.

Will obviously file a PR against the final target at https://github.com/jenkinsci/jep once I feel this is ready for actual review and acceptance there.

I am well aware this might scatter feedback in two different places, but I hope to use this PR to iron out obvious issues, and _officially_ submit upstream only once I can say there's already an agreement the proposal is decent and has been reviewed not only by myself.

Also because as per the process, I consider we are at the ["Initial discussion"](https://github.com/jenkinsci/jep/tree/master/jep/1#jep-workflow) stage, and hence I am not supposed to submit yet.
So I'm _just_ using a GitHub PR just for its review facilities. I could also have used Google docs or something similar for this stage of the process.

So. Feedback welcome :)

### References
* [JENKINS-49406 related JIRA issue](https://issues.jenkins-ci.org/browse/JENKINS-49406)
* [Associated thread on the developers mailing list](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!topic/jenkinsci-dev/XdXuMFLXKPw).
* [Jenkins Essentials' JEP-300](https://github.com/jenkinsci/jep/tree/master/jep/300)

batmat committed:
[JENKINS-49406] WIP Evergreen snapshotting data safety system

WIP

batmat committed:
Precision why compatibleSinceVersion is not used (for now at least)

jtnord commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+
+We plan to configure Jenkins to use non default values to put the workspace and builds not under the same location as job configuration itself.
+
+_Workspace Root Directory_:: `${JENKINS_HOME}/workspace/${ITEM_FULLNAME}`
```
Why do you even have a workspace - executors on the master is evil and should be killed at the earliest convienience?

jglick replied:
[JENKINS-49861](https://issues.jenkins-ci.org/browse/JENKINS-49861) FTR. I plan to have it be possible to leave an executor on the master for use by privileged jobs ([JENKINS-24513](https://issues.jenkins-ci.org/browse/JENKINS-24513)). At any rate, it is worth at least reminding readers that this setting should generally have no effect.

batmat replied:
Was a mistake @jtnord @jglick, I screwed up a copy paste when writing it.

My intent was to put both workspace and builds under the same global root, and under the same item each time. 

Can you have a second look on that specific part?

jtnord commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+
+_Workspace Root Directory_:: `${JENKINS_HOME}/workspace/${ITEM_FULLNAME}`
+_Build Record Root Directory_::
+`${JENKINS_HOME}/var/${ITEM_FULL_NAME}/builds`
```
`var` looks a little weird.   why was this chosen over something more obvious like `builds`
var implies you can stuff more in here - but there is not enough in the path to disambiguate from builds.

jtnord commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+=== Why Git
+
+Using filesystem-level tools offering a snapshotting feature, like LVM, ZFS or btrfs to give a few examples, was considered.
+But this was evacuated because _Essentials_ vision is about providing an link:https://github.com/jenkinsci/jep/tree/71d9391744c8cc7d6595805f7fdd327eedf6811a/jep/300#automatically-updated-distribution["_easier to use_ and _easier to manage_ Jenkins environment"].
```
`evacuated`?  discounted?

batmat replied:
Fixed.

jtnord commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+
+=== Man In The Middle
+
+The main issue here is that an attacker could for instance instruct the evergreen client to ignore everything (by putting `*` in `.gitignore`), hence make it impossible to rollback.
```
don;t use .gitignore then.  use `-f` to add?
move builds outside JENKINS_HOME  (have a directly at the mount that is /builds /whatever /config  and make "config JENKINS_HOME where the usual stuff is stored).

batmat replied:
> don;t use .gitignore then. use -f to add?

Not convinced this fondamentally changes the issue. The goal is still to be able to upgrade the evergreen-client too. So, be it by hacking the `.gitignore` content, or by hacking what things are passed to `-f` I guess the issue is to really find robust ways to not accept instructions from some wrong _instructors_.

 > move builds outside JENKINS_HOME (have a directly at the mount that is /builds /whatever /config and make "config JENKINS_HOME where the usual stuff is stored).

Good point. I gave it some thoughts, but kept it under JENKINS_HOME to _respect_/be consistent with JEP-301:

>  [it is expected that all mutable state and data will be written and stored in JENKINS_HOME, including: Data generated by the Jenkins process. ...](https://github.com/jenkinsci/jep/tree/master/jep/301#mutable-data)

jtnord replied:
just because that is written does not mean it is correct - infact that section of JEP is bad bad bad...

why? 
--pluginRoot --webroot.   JENKINS_HOME is possibly mounted on a resilient storage (and is usually slower or more costly)  there is no need to unpack the Jenkins plugins or war into jenkins-home...

Also logs... Hmmm if they are in jenkins home and you rollback then you loose some logs..
that;s not a good thing is it?

and well if you take it at face value...
> "No other data should be written to the filesystem outside of JENKINS_HOME."

build artifacts are data... yet with evergreen you want to move these to S3 or somewhere not the filesystem.  that JEP would preclude that work.  //cc @rtyler

now if jep-301 wanted to say thal all data be contained under a single directory then fine (but JENKINS_HOME does not need to be at the root of that directory) IMO and if you are making a more complex system because of that descision then go revisit it...

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+= JEP-302: Evergreen snapshotting data safety system
```
You do not pick the number when submitting a JEP. See the instructions.

batmat replied:
Fixed thanks. Didn't rename the folder though to avoid GitHub folding **all** comments below despite me not having addressed them yet. Will do when submitting upstream.

jglick replied:
I think GitHub is smart enough to handle rename detection, but I can understand not wanting to test that now!

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+7. [[healthcheck]] Check it is responding in a healthy manner.
+8. If not:
+.. Take a snapshot footnote:[this way, if new files were created, we don't just delete them in an unrecoverable way when going back to the previous snapshot].
+.. Rollback to the previous data snapshot and Essentials BOM version.
```
_Roll back_

batmat replied:
Fixed.

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+8. If not:
+.. Take a snapshot footnote:[this way, if new files were created, we don't just delete them in an unrecoverable way when going back to the previous snapshot].
+.. Rollback to the previous data snapshot and Essentials BOM version.
+To do so, we will generally create a new commit (i.e. avoid `git reset --hard HEAD~`), to keep a durable track of where we went through, accessible through `git log`.
```
So, `git revert`?

batmat replied:
Yes, basically.

**UPDATE: fixed**

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+To do so, we will generally create a new commit (i.e. avoid `git reset --hard HEAD~`), to keep a durable track of where we went through, accessible through `git log`.
+.. Start (previous) Jenkins version
+// what if starting the previous version doesn't work either?
+9. Report the outcome to the Essentials backend.
```
i.e., let the heads start rolling

batmat replied:
Well, I hope this is clear, but here I meant that we want to report the outcome, whatever it is, to the backend. We also very much need the result if everything goes well, to be able to push it to more instances.

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+
+* Update `.gitignore` content with current Essentials release.
+* `git add .`
+* `git add -u `
```
stray space breaks formatting

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+* Update `.gitignore` content with current Essentials release.
+* `git add .`
+* `git add -u `
+* git commit
```
backticks

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+
+* Update `.gitignore` content with current Essentials release.
+* `git add .`
+* `git add -u `
```
What is the purpose of this command (`--update`)?

> Update the index […]. This removes as well as modifies index entries to match the working tree, but adds no new files.
> If no <pathspec> is given when `-u` option is used, all tracked files in the entire working tree are updated […].

Did you mean to use `-A` / `--all`? (in which case `git add .` is superfluous I guess)

batmat replied:
Yes, I mixed things up here. I was sure `git add .` didn't handle deleted files (wondering if this changed at some point... ?!), so that's why I did two lines. 

In any case, fixed to use `git add --all` which also makes the intent clearer that we want _all_ files, added, modified, or removed.

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+////
+====
+
+===== What to backup
```
`back up`

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+In other words, what is going to contain `.gitignore`.
+As said above, this will likely be improved as we go.
+
+.\.gitignore
```
?

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+_Build Record Root Directory_::
+`${JENKINS_HOME}/var/${ITEM_FULL_NAME}/builds`
+
+Though we will for now backup these directories too, having this distinction baked in from the ground up could prove useful in the future if we need to make the snapshotting and rollback processes smarter and more robust.
```
I think you should not back up build records. They are not configuration and this is not scalable. Just move them to some folder outside `$JENKINS_HOME`.

If we have to roll back software, there is a small chance that build records created between the update and the rollback will no longer be readable in part or at all (since you do not seem to be paying any attention to `compatibleSinceVersion`). However, this should not be fatal—there would be some warnings under **Manage Old Data** and in extreme cases the entire build might go missing in the build history list (probably to reappear after the next successful update). Far more often, there would have been no affect on `build.xml` format, and so reverting the snapshot would just cause a perfectly good build to mysteriously disappear, even though its artifacts may be out in the wild.

In fact many _job_ definitions are not configuration. See [JENKINS-50173](https://issues.jenkins-ci.org/browse/JENKINS-50173).

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+
+We decided to not do, for now at least, for two reasons:
+
+* this looks like an _optimization_. $
```
formatting

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+TIP: A given plugin can indicate what is "the oldest version [...] configuration-compatible with.".
+In other words, if I am changing the configuration format of a plugin in non backwards compatible manner, I am _expected_ to express it as such using link:https://jenkinsci.github.io/maven-hpi-plugin/hpi-mojo.html#compatibleSinceVersion[`compatibleSinceVersion`].
+
+So, this means we could use this field to _optimize_ the way we roll back: once a configuration file has been upgraded by a plugin update, we can exclude it from rollback if the `compatibleSinceVersion` field was not changed.
```
Not following how you got there.

What `compatibleSinceVersion` means is this: if this flag is marked with a value of, say, `1.5`, and you have just updated the plugin from version 1.4 to 1.5, but you wish to revert back to 1.4 after seeing some problems, and you can detect that there is an XML file written by that plugin (visible via the `plugin` attribute):

* You _may_ revert that file to the previous version. By design, Essentials always will (modulo the discussion about `build.xml` and similar build record files).
* You _must not_ leave it at the version written by 1.5. So if you are invested in the modifications to this file made recently, you are in trouble.

But you seem to be talking about the opposite: the usual case that the upgrade is _not_ crossing a `compatibleSinceVersion` boundary. In that case it is fine to let the 1.4 plugin read a modification to the config file. But that is irrelevant for this JEP since you are versioning all of `$JENKINS_HOME` just in case.

batmat replied:
I think my phrasing might have been a bit convoluted, but I think we are saying the same thing.

I wrote, chosen excerpt:
> exclude it from rollback if the `compatibleSinceVersion` field was not changed

I.e. taking your example to illustrate: when going from version `1.4` to `1.5`, *but* with `compatibleSinceVersion` staying at `1.4` or lower, then we _could_ avoid rolling the generated config files for this plugin, even if config files with `... plugin="some-feature@1.5"` present.

batmat replied:
Heavily rephrased in last commit.

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+
+* this looks like an _optimization_. $
+Hence as such, this is probably premature to try and be very smart with the way the downgrade will work ;
+* there is no guarantee that all plugin developers currently know and think about updating this metadata when needed.
```
There is no guarantee that all plugin developers know anything at all. We are talking here about developers working on plugins included in Essentials. These developers are expected to follow proper Jenkins plugin development practices, which included using `compatibleSinceVersion` when making format changes which would prevent an older version of the plugin from gracefully reading data written by a newer version.

(Generally it is OK to just add fields and the like—the older version would not know what to do with these anyway. The cases where this is needed are things like moving data around; SCM API 2.0 in the extreme case, but could be something as small as `readResolve`ing a `String count` field to an `int counter` and deprecating the former.)

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+* this looks like an _optimization_. $
+Hence as such, this is probably premature to try and be very smart with the way the downgrade will work ;
+* there is no guarantee that all plugin developers currently know and think about updating this metadata when needed.
+This is likely something that will be improveable when we work on link:https://issues.jenkins-ci.org/browse/JENKINS-49806[JENKINS-49806] and write the associated Jenkins Enhancement Proposal.
```
sp.

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+
+== Backwards Compatibility
+
+There are no backwards compatibility concerns related to this proposal.
```
Well, other than the entire proposal hinging on subtleties of compatibility. :-)

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+=== Secrets
+
+Versioning secrets is not an issue, as the snapshotting system is going to be local to the running instance.
+As the Git repository data will never be pushed _outside_, no data leak is expected from this side.
```
I do not think this is acceptable. Someone will use the Git repository externally regardless of what you advertise.

Fortunately there is a simple solution: add

```
secrets/master.key
```

to `.gitignore`. It is created on first startup and never changed thereafter, so it would never be affected by an upgrade or downgrade. And without it, all other secrets remain (theoretically) secret.

[Reference](https://go.cloudbees.com/docs/cloudbees-documentation/cje-user-guide/index.html#backup-sect-configure)

batmat replied:
Well, I do not feel strongly to add it or ignore it. But well, if someone is able to access the FS and so what you are saying, in any case they have the power to read that key.

I understand what you are saying as "someone could have the bad idea to push that repo elsewhere, when it's not supposed to be, not groking the fact they're going to leak central secrets doing so, and so we'll be seen as responsible for it.".

So yes, let's be conservative and not encourage people to shoot themselves in the foot.

batmat replied:
FTR, fixed.

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+
+* Apply the upgrade or downgrade, then check the instance is _running fine_:
+** No administrative monitor triggered
+** No new ERROR logs showing up
```
You mean, `Level.ERROR`? Hardly useful: few plugins log at this level to begin with. Almost everything of interest would be a `WARNING`.

batmat replied:
Yes and no. But your point makes me realize I really need to _externalize_ this concept. We need a standalone design for determining Jenkins health I think. It's a critical condition/piece for many components: for testing, for evergreen-client to decide to rollback or not, etc. And it is not (necessarily) related to the data snapshotting system.

So will refactor to make this generic and point to something to be done in this regard.

Also commented in https://issues.jenkins-ci.org/browse/JENKINS-49406?focusedCommentId=332233&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-332233 about that

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+** No administrative monitor triggered
+** No new ERROR logs showing up
+
+We will need to develop adhoc testing tools to be able to automatically assess the health of a Jenkins Essentials instance after an upgrade or a downgrade.
```
_ad-hoc_

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+
+Before delivering updates on real connected instances, we must test at least the following scenarios.
+
+* Apply the upgrade or downgrade, then check the instance is _running fine_:
```
I think you are missing something critical here: you need to be able to ensure that nothing interesting happening in the system between the time Jenkins completes initialization after the upgrade, and the time the health check completes. If builds run, that is probably fine (as I argue above w.r.t. build record location), though a `QueueTaskDispatcher` or just `setQuietDown` could block builds temporarily after startup. But what about users editing job or system configuration? You would need to put some kind of UI block in place, I think.

The broader question which I guess you are punting on is how useful a simplistic health check like this really is. Just starting without warnings is pretty useless if every build you try to run bombs out with some fatal low-level error. But then if you let some trial builds run, you are widening the window after which a revert could be problematic.

jglick commented on jep/302/README.adoc:
```diff
@@ -0,0 +1,260 @@
+Detailing this here is out of scope for this proposal.
+It is however highly desirable that we centralize this logic and use it both during automated tests, and in production for the evergreen-client to automatically analyze if a product instance is healthy or is not (and decide to rollback or not, for the current matter here).
+
+We will leverage the link:https://github.com/cloudbees/operations-center-acceptance-test/[Jenkins Acceptance Test Harness project] for this purpose.
```
I think you meant to link to something else.

batmat replied:
oooops fixed

batmat committed:
Back to JEP-0000

I was not supposed to choose the number myself, obviously...

Thanks Jesse for the reminder.

Not renaming the folder yet, on purpose, to avoid folding all other
review comments although I didn't address them yet.

batmat committed:
rollback must be roll back when used as a verb

batmat committed:
Add secrets/master.key to .gitignore

Jesse is basically saying that despite the system/git repo is
not designed to be pushed outside, (some) people will do so...

batmat committed:
Woops, countersense

batmat committed:
Small formatting issues

batmat committed:
Clarify that `revert` is meant

batmat committed:
Use git add --all instead of a convoluted git add . + git add -u

batmat committed:
back up, not backup, when used as a verb

batmat committed:
Formatting issue

Was trying to use the title of a code block, starting with a dot.
But here asciidoctor was confused by the .. at ..gitignore.

So using the clearer form title=blah to make it work
(and be less magic in the same go, which is good).

batmat committed:
Sigh, critical copy-paste error...

The intent here was to have a single root for builds and
workspace for a given item.

batmat committed:
evacuated => discounted

batmat committed:
Clarify a bit more the role of `${JENKINS_HOME}/var/`

batmat committed:
adhoc => ad-hoc

batmat committed:
spurious char

batmat committed:
typo

batmat committed:
Wrong link...

batmat commented:
Reminder for me, and for anyone interested, things I want to address:
1. About the point above by @jglick and @jtnord on moving `$JENKINS_HOME/var` out of `$JENKINS_HOME`, [I opened a dedicated thread for feedback on the dev list](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!topic/jenkinsci-dev/xiaHpfGPTZ8). I am leaning towards I'm going to adjust the proposal with this. It makes sense to me.
2. [Another important point above](https://github.com/batmat/jep/pull/1#discussion_r174518196) by @jglick which I'm summarizing as "how do we check Jenkins is *actually* healthy **before** opening the flood gates to users: letting jobs execute, people modify things through the UI, etc.". This point is probably the most critical and needs addressing.

If there is anything else important I missed, please don't hesitate to tell me so.

batmat committed:
Data snapshotting system is NOT a backup system

batmat committed:
Fix wrong title level

batmat committed:
Adjust the data segregation part of the proposal

batmat committed:
Clarify that determining the health of an instance is to be done, separately

batmat committed:
Progressive startup

Proposal to address Jesse's comment:

> The broader question which I guess you are punting
> on is how useful a simplistic health check like this
> really is. Just starting without warnings is pretty
> useless if every build you try to run bombs out with
> some fatal low-level error. But then if you let some
> trial builds run, you are widening the window after
> which a revert could be problematic."

batmat committed:
Heavily rephrase the compatibleSinceVersion section

batmat committed:
Final adjustments, leaving the health checking out of scope

as it should have been since the beginning

jglick commented on jep/302/README.adoc:
```diff
@@ -100,17 +99,21 @@ To do so, we will generally create a new commit (i.e. avoid `git reset --hard HE
+The intent is to have everything not ignored in the index.
+In other words, `git status` just after a SNAPSHOT should be in the `nothing to commit, working tree clean` state.
+
+We will *not* snapshot data that is too big or does not make sense to store:.
```
stray `:`

batmat replied:
Fixed.

jglick commented on jep/302/README.adoc:
```diff
@@ -122,37 +125,81 @@ We need to finish up the work on the BOM to be more precise here.
+│   ├── nodes
+│   ├── plugins
+│   ├── secrets
+│   ├── updates
```
can be ignored I think

batmat replied:
Added to `.gitignore`.

jglick commented on jep/302/README.adoc:
```diff
@@ -122,37 +125,81 @@ We need to finish up the work on the BOM to be more precise here.
+
+This way, everything related to jobs, but is less important than its configuration itself is separated under a different `${JENKINS_HOME}/var/` root.
+
+Though we will for now back up these directories too, having this distinction baked in from the ground up could prove useful in the future if we need to make the snapshotting and rollback processes smarter and more robust.
```
Wait, I thought we were only backing up `/var/jenkins/home`?

batmat replied:
yes, that's it. Reading, maybe I forgot to adjust a sentence.

batmat replied:
Fixed, removed that part.

jglick commented on jep/302/README.adoc:
```diff
@@ -122,37 +125,81 @@ We need to finish up the work on the BOM to be more precise here.
+====== about `/var/jenkins/home/plugins`
+
+This directory contains the hpi/jpi files before extraction.
+Ideally, we should be moving this elsewhere under `/var/jenkins/var/plugins`, but it's not doable yet currently (`--pluginsroot` only configures a different location for exploded plugins).
```
Careful here. Currently `$JENKINS_HOME/plugins/shortName.jpi` is the definitive record that a plugin was installed (there is also a `shortName.jpi.disabled` file if disabled), and in what version. If you do not back that up, then you are no longer versioning the Jenkins configuration.

Now I understand that for plugins in Essentials, this is exactly right—the Essentials manifest is in control of what plugins are there. (Again, excepting `*.jpi.disabled` files, so I think what you meant to exclude is more narrowly `plugins/*.jpi`.) But what about “inessential” plugins? There will be no trace in the Git history of their presence or version.

Admittedly a rollback is not going to _need_ to revert these, either, since the Essentials upgrade would not have touched them. Just seems a little risky. People may be expecting that a Git revision (plus `master.key`) corresponds to a fully reconstructible Jenkins instance.

jglick replied:
Maybe the `.gitignore` can simply list those plugins actually in Essentials?

```
plugins/structs.jpi
# …
```

batmat replied:
> People may be expecting that a Git revision (plus master.key) corresponds to a fully reconstructible Jenkins instance.

Well, then they will be deceived... It's not a backup system as we already discussed.

To me that this part is going to be out of the scope of the snapshotting system, exactly for the reason you gave: I consider that the whole plugin list is defined explicitly by the BOM, for Essentials plugin.

I would say that for others, well it's also out of scope, and it will be the evergreen client responsibility to store the versions of **all** plugins present there, to be able to restore the previous version if we rollback.

jglick commented on jep/302/README.adoc:
```diff
@@ -122,37 +125,81 @@ We need to finish up the work on the BOM to be more precise here.
+=== Checking Jenkins health
+
+From the perspective of this proposal, this is out of scope.
+But the outter _controller_ of the upgrade, the evergreen client, will need a way to decide if a rollback must be triggered or not.
```
typo

batmat replied:
fixed

jglick commented on jep/302/README.adoc:
```diff
@@ -163,32 +210,58 @@ In the context of _Essentials_, we cannot rely on external backups to revert to
+If a version is proved to be problematic after a few days, the data snapshotting system will **not** be used.
+
+This would be quite impractical because the instance probably generated actual work items during this timeframe.
+So rolling back that much later would be subject to losing data.
```
grammar: _would risk data loss_

jglick replied:
also: _rolling that back_ (separable verbs, sigh)

batmat replied:
done thanks

jglick commented on jep/302/README.adoc:
```diff
@@ -198,12 +271,14 @@ There are no backwards compatibility concerns related to this proposal.
+Versioning secrets should not be an issue per se, as the data snapshotting system is designed to be local to the running instance.
+In other words, the Git repository data will never be pushed _outside_ by the _Essentials_ code, so no data leak is normally expected from this side.
+
+But as users may have the unfortunate idea to push that repository elsewhere, not being aware they are leaking all secrets, we will conservatively add `secrets/master.key` to the `.gitignore` file.
```
Would be a good place to briefly explain what this file is and why excluding it from the Git repository prevents any straightforward extraction of secrets.

batmat replied:
I would like to avoid (badly) repeating something that probably ought to be defined in the official docs. But can add a link. 

Something like I guess:

> "``Secret``s are encrypted using the `master.key` files, so we really want to avoid attackers to be able to access both `credentials.xml` and `master.key` at the same time

?

batmat committed:
Remove leftover from previous versions

batmat committed:
outter => outer

batmat committed:
better grammar, thanks Jesse

batmat commented:
Now submitted for official review at https://github.com/jenkinsci/jep/pull/67
